### PR TITLE
Fix reencrypt routes and add path based support

### PR DESCRIPTION
**Pull request by ramr**
Get reencrypt routes working + add support for path based reencrypt routes - similar to the other http/edge path based checks. This is the same functionality #4097 was going to fix.
This fixes issue #4011

Whilst testing this, I did find some issues with ACLs (```map_beg``` doesn't seem to work inside an ACL), so I inlined the ACL code to get it to work and then it further got optimized because the check/condition in  
```use_backend be_secure_%[base,map_beg(/var/lib/haproxy/conf/os_reencrypt.map)] if { base,map_beg(/var/lib/haproxy/conf/os_reencrypt.map) -m found }``` was redundant and so it morphed into its current form. LMK if that's outta whack - it does work fine.

Test with:  
```
function test_uris() {  
   local verbosity=""   #  "-vvv"  
   local rip="10.245.2.3"  # "10.0.2.15"
   for uri in https://www.example2.com/test/ https://www.example2.com/; do   
     echo "uri: $uri -> $(curl -s -w "%{http_code}" --resolve www.example2.com:443:$rip --cacert _qatests/hello-nginx-docker/certs/mypersonalca/certs/ca.pem $verbosity "$uri")"  
  done  
}   
  
oc create -f https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/routing/nginx-pod.json  
oc create -f https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/routing/reencrypt/service_secure.json  
test_uris  
  
oc create -f https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/routing/reencrypt/route_reencrypt-path.json  
test_uris  
  
oc create -f https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/routing/reencrypt/route_reencrypt.json  
test_uris  
  
oc delete -f https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/routing/reencrypt/route_reencrypt-path.json  
test_uris  
```

@pweil- @rajatchopra  PTAL Thx

### CODE REVIEW

**Comment by pweil-** [comment]
Load Balancer Weekly declared this PR "a shining example dynamic routing".  

Can we remove the `if TRUE` from all the checks and clean the template up even more?  Also, would it be easier to just iterate through `*.map` in the reload script.  Then you wouldn't have to touch that thing again if anything gets added later - not that I see that in the future but just a thought.

**Comment by ramr** [comment]
Yup, that's a good idea ... never say never though!! ;^) 
Done - cleaned up the ```IF TRUE``` + iterated over the glob pattern ```$configdir/*.map```.

**Comment by pweil-** [comment]
cool.  Looks good.  @rajatchopra you want to take a second pass?

**Comment by rajatchopra** [comment]
LGTM
[merge]

**Comment by ramr** [comment]
Oh, boy this was a doozy -- @rajatchopra and I were debugging the test case failures ... and in that we saw some issues + inefficiencies, so just cleaned all that up + squashed into a single new commit and added re-encrypt tests. The new acls/conditions don't use the hdr(host) anymore because ```base```  matched to ```map_beg``` will catch all those - longest (full path match) to shortest (just host matches). 
Just fyi, whilst debugging this, we found that it was never reaching the ```hdr(host)``` rule because without an acl/condition haproxy tries to find a backend from the map and if it doesn't find one it uses the ```default_backend``` with ```use_backend```, all subsequent ```use_backend``` checks are ignored.
Anyway, long story short, its fixed with this change + has tests and all router-specific tests pass locally.

@pweil-  @rajatchopra  PTAL  Thx

**Comment by pravisankar** [comment]
[test]

**Comment by ramr** [comment]
[test]

File: images/router/haproxy/conf/haproxy-config.template [hunk 1]
```diff
@@ -84,16 +84,18 @@ frontend fe_sni
...
   mode http
 
-  # re-ssl?
-  acl reencrypt hdr(host),map(/var/lib/haproxy/conf/os_reencrypt.map) -m found
-  use_backend be_secure_%[hdr(host),map(/var/lib/haproxy/conf/os_tcp_be.map)] if reencrypt
+  # check re-encrypt backends first - from most specific to general path.
+  acl reencrypt base,map_beg(/var/lib/haproxy/conf/os_reencrypt.map) -m found
+
+  # Search from most specific to general path (host case).
+  use_backend be_secure_%[base,map_beg(/var/lib/haproxy/conf/os_reencrypt.map)] if reencrypt
```
Comment (rajatchopra) [inline images/router/haproxy/conf/haproxy-config.template]: This means a double lookup, right? whats wrong with if TRUE?

Comment (ramr) [inline images/router/haproxy/conf/haproxy-config.template]: Yeah, it does mean a double lookup only in the case its in the re-encrypt map. 
Don't think ```if TRUE``` will work because the ```use_backend``` is only meant processed if the condition evaluates to true. If we always do it for ```if TRUE```, we get back to the other behavior of falling back to using the ```default_backend``` if its not found.

**Comment by rajatchopra** [comment]
[merge]
